hal/src/common/dmac/channel.rs

```diff
@@ -0,0 +1,313 @@
+
+/// These methods may be used on any DMA channel in any configuration
+impl<S, const ID: u8> Channel<S, ID> {
+    /// Set channel ID and run the closure. A closure is needed to ensure
+    /// the registers are accessed in an interrupt-safe way, as the SAMD21
+    /// DMAC is a little funky - It requires setting the channel number in
+    /// the CHID register, then access the channel control registers.
+    /// If an interrupt were to change the CHID register, we would be faced
+    /// with undefined behaviour.
+    #[cfg(any(feature = "samd11", feature = "samd21"))]
+    fn with_chid<F: Fn(&mut DMAC)>(&mut self, dmac: &mut DMAC, fun: F) {
+        cortex_m::interrupt::free(|_| {
+            // SAFETY: this is actually safe as long as we write a correct channel number to
+            // the CHID register
+            unsafe {
+                dmac.chid.modify(|_, w| w.id().bits(ID));
+            }
+
+            fun(dmac);
+        });
+    }
+
+    /// Set channel ID and run the closure. A closure is needed to ensure
+    /// the registers are accessed in an interrupt-safe way, as the SAMD21
+    /// DMAC is a little funky. For the SAMD51/SAMEx, we simply take a reference
+    /// to the correct channel number and run the closure on that.
+    #[cfg(any(
+        feature = "samd51",
+        feature = "same51",
+        feature = "same53",
+        feature = "same54"
+    ))]
+    fn with_chid<F: Fn(&CHANNEL)>(&mut self, dmac: &mut DMAC, fun: F) {
+        let mut ch = &dmac.channel[ID as usize];
+        fun(&mut ch);
+    }
+
+    /// Configure the DMA channel so that it is ready to be used by a
+    /// [`DmaTransfer`](super::transfer::DmaTransfer).
+    ///
+    /// # Return
+    ///
+    /// A `Channel` with a `Ready` status
+    pub fn init(
+        mut self,
+        lvl: PriorityLevel,
+        enable_interrupts: bool,
+        controller: &mut DmaController,
+    ) -> Channel<Ready, ID> {
+        let dmac = controller.as_mut();
+
+        // Software reset the channel for good measure
+        self._reset_private(dmac);
+
+        self.with_chid(dmac, |d| {
+            // Reset the channel to its startup state and wait for reset to complete
+            d.chctrla.modify(|_, w| w.swrst().set_bit());
+            while d.chctrla.read().swrst().bit_is_set() {}
+
+            #[cfg(any(feature = "samd11", feature = "samd21"))]
+            // Setup priority level
+            d.chctrlb.modify(|_, w| w.lvl().bits(lvl as u8));
+
+            #[cfg(any(
+                feature = "samd51",
+                feature = "same51",
+                feature = "same53",
+                feature = "same54"
+            ))]
+            d.chprilvl.modify(|_, w| w.prilvl().bits(lvl as u8));
+
+            if enable_interrupts {
+                // Enable all interrupt sources
+                d.chintenset
+                    .modify(|_, w| w.susp().set_bit().tcmpl().set_bit().terr().set_bit());
```

Comment:
I would avoid chaining writes like this. `rustfmt` does terrible things, and this is really unreadable. I would stick with multiple calls of `w.field().change();`. You'll just need to leave the semi-colon off the last one.